### PR TITLE
iOS: Expose builldBeacon and cancelBeacon hooks inside of BaseBeaconPlugin and make BeaconPlugin open class

### PR DESCRIPTION
**Pull request by sakuntala-motukuri**
### Change Type (required)
Indicate the type of change your pull request is:

iOS: Expose builldBeacon and cancelBeacon hooks inside of BaseBeaconPlugin and make BeaconPlugin open class

-  Changed return type of tap in Hook.swift to generic instead of void
- Updated cancelBeacon and buildBeacon hooks accordingly in BaseBeaconPlugin class
- Made BeaconPlugin open to extend the access for extending classes
- Updated tests for above

- [ ] `patch`
- [x] `minor`
- [ ] `major`

### Does your PR have any documentation updates?
- [ ] Updated docs
- [x] No Update needed
- [ ] Unable to update docs

<details>
  <summary>📦 Published PR as canary version: <code>0.9.2--canary.536.18200</code></summary>
  <br />

  Try this version out locally by upgrading relevant packages to 0.9.2--canary.536.18200
</details>

### CODE REVIEW

File: plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift [hunk 1]
```diff
@@ -136,4 +136,105 @@ class BaseBeaconPluginTests: XCTestCase {
...
+        ))
+        
+        guard let hooks = plugin.hooks else {
+            XCTFail("Hooks are not initialized")
+            return
+        }
+        
+        hooks.buildBeacon.tap { (arg1: JSValue, arg2: JSValue) -> JSValue? in
+            buildBeaconHandler.fulfill()
+            if let actionDict = arg1.toDictionary(),
```
Comment (nancywu1) [inline plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift:223]: can we do something in the buildBeacon like modify a value in the beacon like the `data` similar to https://github.com/player-ui/player/blob/main/plugins/beacon/core/src/index.test.ts#L136 and then verify in the handler the modified value?

File: plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift [hunk 2]
```diff
@@ -136,4 +136,105 @@ class BaseBeaconPluginTests: XCTestCase {
...
+        let context = JSContext()!
+        JSUtilities.polyfill(context)
+        let buildBeaconHandler = expectation(description: "Build Beacon Handler called")
+        let handlerExpectation = expectation(description: "Handler called")
+        let handler = MockHandler()
+        
+        let plugin = BaseBeaconPlugin<DefaultBeacon>(plugins: []) { (beacon) in
+            if handler.calls.isEmpty {
+                handler.handle(beacon.viewId!, beacon.data)
+                handlerExpectation.fulfill()
```
Comment (nancywu1) [inline plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift:202]: does this ever get called?

File: plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift [hunk 3]
```diff
@@ -136,4 +136,105 @@ class BaseBeaconPluginTests: XCTestCase {
...
+            return false
+        }
+        
+        wait(for: [handlerExpectation, cancelBeaconHandler], timeout: 10)
+    }
+    
+    class MockHandler {
+        var calls: [(String, Any?)] = []
+        
+        func handle(_ action: String, _ data: Any? = nil) {
```
Comment (nancywu1) [inline plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift:186]: i dont think this MockHandler is needed if we can test with expectations

File: plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift [hunk 4]
```diff
@@ -136,4 +136,105 @@ class BaseBeaconPluginTests: XCTestCase {
...
+            data: .string(data: "example")
+        ))
+        
+        guard let hooks = plugin.hooks else {
+            XCTFail("Hooks are not initialized")
+            return
+        }
+        
+        hooks.cancelBeacon.tap { (arg1: JSValue, arg2: JSValue) -> Bool in
+            cancelBeaconHandler.fulfill()
```
Comment (hborawski) [inline plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift:173]: I'm surprised this is working, the hook should get called by the underlying JS plugin when `beacon` is called, but you are tapping after calling `beacon`?

File: plugins/beacon/ios/Sources/BaseBeaconPlugin.swift [hunk 5]
```diff
@@ -94,23 +107,35 @@ open class BaseBeaconPlugin<BeaconStruct: Decodable>: JSBasePlugin {
...
+                let beacon = try AnyTypeDecodingContext(rawData: data)
                     .inject(to: JSONDecoder())
                     .decode(BeaconStruct.self, from: data)
-            else { return }
-            self?.callback?(beacon)
+                
+                self?.callback?(beacon)
+            } catch {
+                print("Error during JSON serialization or deserialization: \(error)")
+            }
```
Comment (hborawski) [inline plugins/beacon/ios/Sources/BaseBeaconPlugin.swift:128]: This section should be reverted, we should not use `print` statements in code we publish

**Review by hborawski — CHANGES_REQUESTED** [review]

File: plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift (hunk 2 above)
Comment (sakuntala-motukuri) [reply plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift:202]: This piece of line sets up the buildBeacon hook and fulfills the buildBeaconHandler expectation when the hook is triggered.

File: plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift (hunk 3 above)
Comment (sakuntala-motukuri) [reply plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift:186]: I'm using MockHandler to test this piece of code https://github.com/player-ui/player/pull/536/files#diff-f6cdb5dd4b4782a5f8b61d122af9dc5c90dad5cdbe669551df7e4e9ce855b37fR201

File: plugins/beacon/ios/Sources/BaseBeaconPlugin.swift (hunk 5 above)
Comment (sakuntala-motukuri) [reply plugins/beacon/ios/Sources/BaseBeaconPlugin.swift:128]: Do you want me to remove all the print statements ? as I'm using some for debugging errors during the serialization process if any like catch {
                print("Error during JSON serialization or deserialization: \(error)")
            }

File: plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift (hunk 4 above)
Comment (sakuntala-motukuri) [reply plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift:169]: fixed

File: plugins/beacon/ios/Sources/BaseBeaconPlugin.swift (hunk 5 above)
Comment (sakuntala-motukuri) [reply plugins/beacon/ios/Sources/BaseBeaconPlugin.swift]: fixed

File: plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift (hunk 2 above)
Comment (sakuntala-motukuri) [reply plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift]: fixed

File: plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift (hunk 1 above)
Comment (sakuntala-motukuri) [reply plugins/beacon/ios/Tests/BaseBeaconPluginTests.swift]: fixed

File: plugins/beacon/swiftui/Sources/SwiftUIBeaconPlugin.swift [hunk 6]
```diff
@@ -18,20 +18,14 @@ import PlayerUIBaseBeaconPlugin
 /**
  Plugin used by `SwiftUIPlayer` for beaconing in a uniform format between platforms
  */
-public class BeaconPlugin<BeaconStruct: Decodable>: BaseBeaconPlugin<BeaconStruct>, NativePlugin {
-    /**
```
Comment (nancywu1) [inline plugins/beacon/swiftui/Sources/SwiftUIBeaconPlugin.swift:22]: looks like comments were removed by accident?

**Review by nancywu1 — APPROVED** [review]

**Pull request merged**